common/djangoapps/student/models.py

```diff
@@ -1552,11 +1553,163 @@ def send_signal_full(cls, event, user=user, mode=mode, course_id=None, cost=None
+            timezone_offset = get_time_zone_offset(user_timezone, block_date)
+            block_date = block_date.astimezone(user_timezone)
+
+            if block_date.date() == course_date_list[0]['date']:
+                if course_date_list[0]['assignment_count'] == -1:
+                    course_date_list[0].update({
+                        'title': block['title'] or '',
+                        'assignment_type': block['assignment_type'] or '',
+                        'link': block['link'] or '',
+                        'assignment_count': (course_date_list[0]['assignment_count'] + 1 if block['assignment_type']
+                                             else course_date_list[0]['assignment_count']),
+                        'due_date': block_date.strftime("%a, %b %d, %Y"),
+                        'due_time': (f'{block_date.strftime("%H:%M %p")} GMT{timezone_offset}'
+                                     if block['assignment_type'] else '')
+                    })
+                else:
+                    course_date_list[0]['assignment_count'] += 1
+            elif block_date.date() == course_date_list[2]['date']:
+                if course_date_list[2]['assignment_count'] == -1:
+                    course_date_list[2].update({
+                        'title': block['title'] or '',
+                        'assignment_type': block['assignment_type'] or '',
+                        'link': block['link'] or '',
+                        'assignment_count': (course_date_list[2]['assignment_count'] + 1 if block['assignment_type']
+                                             else course_date_list[2]['assignment_count']),
+                        'due_date': block_date.strftime("%a, %b %d, %Y"),
+                        'due_time': (f'{block_date.strftime("%H:%M %p")} GMT{timezone_offset}'
+                                     if block['assignment_type'] else '')
+                    })
+                else:
+                    course_date_list[2]['assignment_count'] += 1
+            elif block_date.date() == course_date_list[1]['date']:
+                if course_date_list[1]['assignment_count'] == -1:
+                    course_date_list[1].update({
+                        'title': block['title'] or '',
+                        'assignment_type': block['assignment_type'] or '',
+                        'link': block['link'] or '',
+                        'assignment_count': (course_date_list[1]['assignment_count'] + 1 if block['assignment_type']
+                                             else course_date_list[1]['assignment_count']),
+                        'due_date': block_date.strftime("%a, %b %d, %Y"),
+                        'due_time': (f'{block_date.strftime("%H:%M %p")} GMT{timezone_offset}'
+                                     if block['assignment_type'] else '')
+                    })
+                else:
+                    course_date_list[1]['assignment_count'] += 1
```

Comment:
if we define a function, we can simply use that here. this will avoid repetition